### ISSUE TITLE
Compose video recordings using FFMPEG

### ISSUE THREAD

**Issue by JamesPHoughton** [Watts-Lab/deliberation-video-pipeline#2]
To make it easy to double-check that participants are behaving the way we expect, we want a simple way to take the raw track video files and compose them into something that is easy for a human being to view and double check. 

One way to do this is to use FFMPEG to stitch all the video and audio files together and create a new video file that contains all video and audio tracks. In the end we should have something that looks like:
<img width="858" alt="image" src="https://github.com/Watts-Lab/deliberation-video-pipeline/assets/4304478/94374fad-0a17-45aa-bb08-e13adc55cc0f">

clearly it won't have the names, etc. but should be able to see all the participants at the same time and hear all the audio.

# steps
1. work out a correct ffmpeg command that will stitch a demo set of videos together.
2. write a python script, etc. that will take a folder with videos in it, and compose the ffmpeg command properly.

**Comment by mxumary** [comment]
**TL;DR** ffmpeg commands should look something like this:
First, fill in the silence in each audio track based on the audio start time -- we are using `adelay` which is based in milliseconds. The timestamps are found using the `audio_start_time` when you use `fprobe` to get the group data (sandbox.ipynb in the deliberation video analysis repo). Merge the audio files together.
```
ffmpeg \
-i "files/set_two/1677511696683-67eedcee-98bf-4f05-90f7-9e8ed2ce58cb-cam-audio-1677511704290.webm" \
-i "files/set_two/1677511696683-fc210681-bd73-4bfc-ac8d-1c5c606a9bd0-cam-audio-1677511697662.webm" \
-filter_complex "[0:a]adelay=7659|7659[a0]; [1:a]adelay=946|946[a1]; [a0][a1]amerge=inputs=2[a]" -ac 2 -map "[a]" set_two_output_audio.webm
```
Second, overlay the audio track on top of the videos, where they are side-to-side using the `hstack` filter
```
ffmpeg \
-i "files/set_two/set_two_ffmpeg_resampled_person1.mp4" \
-i "files/set_two/set_two_ffmpeg_resampled_person2.mp4" \
-i "set_two_output_audio.webm" \
-filter_complex "[0:v]scale=1280:720[v0];[1:v]scale=1280:720[v1];[v0][v1]hstack=inputs=2[v]" \
-map "[v]" -map 2:a \
files/set_two/set_two_combined_sepaudio.mp4
```

To do next: test on videos with more than two people, make the code more generalized across any set of audio/video file pairs.

Other notes:
I'm pretty sure that we can just use the mp4 files without audio as well, but you will probably have to specify the `itsoffset` for the video start time. The videos that I used in the `hstack` are ones that already have its audio integrated with the video, and the `-itsoffset` for the audio and video were already accounted for.

Why we used `adelay` instead of `itsoffset` for audio:
What was happening with this video specifically was that webm starts each audio flie at its specified timestamp, but ffmpeg merges the two as if you played the two simultaneously, irrespective of the "true" timestamp.

Suppose that person 1 has an audio start time of 7.9 seconds. Person 2 has an audio start time of 0.9 seconds, but starts talking about 7 seconds letter.

If there is no offset, the webm output looks like this: essentially, both people's audios start at the same time.
![image](https://github.com/Watts-Lab/deliberation-video-pipeline/assets/66391466/80826335-ae95-4b67-81b0-2a5c727265ee)

But if you have `-itsoffset` based on the audio start time, you get this:
![image](https://github.com/Watts-Lab/deliberation-video-pipeline/assets/66391466/7b55690c-cc4a-4131-921a-9de600ad1c2a)

Using `adelay` and filling in silence for each person's audio file, based on the start time, ensures that we're synchronizing correctly.
![image](https://github.com/Watts-Lab/deliberation-video-pipeline/assets/66391466/d45e8a55-2515-4b86-b24d-5529d19d50bf)

Additional notes:
We need to test this method across a variety of video/audio examples. There was one set of audio/video recordings where one person's audio intrinsically did not match with its video (e.g., the audio was out of sync in the beginning, but went into sync in the middle/towards the end) when the individual audio/video was merged. I suspect that this exacerbated the out-of-sync behavior from the other two people when all three people were brought into the same video.
